### PR TITLE
[SPARK-10574] [ML] [MLlib] HashingTF supports MurmurHash3

### PR DESCRIPTION
**Pull request by yanboliang**
## What changes were proposed in this pull request?

As the discussion at [SPARK-10574](https://issues.apache.org/jira/browse/SPARK-10574), `HashingTF` should support MurmurHash3 and make it as the default hash algorithm. We should also expose set/get API for `hashAlgorithm`, then users can choose the hash method.

Note: The problem that `mllib.feature.HashingTF` behaves differently between Scala/Java and Python will be resolved in the followup work.
## How was this patch tested?

unit tests.

cc @jkbradley @MLnick

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala [hunk 1]
```diff
@@ -54,20 +60,43 @@ class HashingTF(val numFeatures: Int) extends Serializable {
...
    */
   @Since("1.1.0")
+  @deprecated("Use indexOfTerm instead. This method will be removed in 2.1.0.", "2.0.0")
   def indexOf(term: Any): Int = Utils.nonNegativeMod(term.##, numFeatures)
 
   /**
+   * Returns the index of the input term by different hash algorithms.
+   */
+  @Since("2.0.0")
+  def indexOfTerm(hashAlgorithm: String)(term: Any): Int = {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala]: Why would the user pass the hash algorithm again here? you can store a reference to a function from Any to Int when the hash algorithm is configured, and not even have to look it up every time.

Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala]: Because users may update ```hashAlgorithm``` by ```setHashAlgorithm``` after it first generated, so we should determinate the algorithm used at here.

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala]: Yes, but it doesn't quite make sense as a public API, since the user has already specified the hashing algorithm. It should just accept the term like `indexOf`. Of course, internally it's necessary to use the right algorithm, but that can be saved off as a function from Any to Int whenever the setter is called to change the algorithm. It's not even necessary to pick the algorithm at runtime on every call.

Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala]: Yes, it's better to save it as a function whenever the setter is called then to update the param.In this PR I just follow the exist ML style. I think all ML params setter need this behavior change and it's better to discuss more, so may be do it in a separate PR?

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala]: I don't think you need a Param. Just another private member. It's not logically a separate parameter.

**Comment by ssimeonov** [comment]
When the "hashing trick" is used in practice, it is important to do things such as monitor, manage or randomize collisions. If that show there are problems with the chosen hashing approach, then one would experiment with the hashing function. All this suggests that a hashing function should be treated as an object with a simple interface, perhaps as simple as `Function1[Any, Int]`. Collision monitoring can then be performed with a decorator with an accumulator. Collision management would be performed by varying the seed or adding salt. Collision randomization would be performed by varying the seed/salt with each run and/or running multiple models in production which are identical expect for the different seed/salt used.

The hashing trick is very important in ML and quite... tricky... to get working well for complex, high-dimension spaces, which Spark is perfect for. An implementation that does not treat the hashing function as a first class object would substantially hinder MLlib's capabilities in practice.

File: mllib/src/main/scala/org/apache/spark/ml/feature/HashingTF.scala [hunk 2]
```diff
@@ -63,7 +63,18 @@ class HashingTF(override val uid: String)
...
 
-  setDefault(numFeatures -> (1 << 18), binary -> false)
+  /**
+   * The hash algorithm used when mapping term to integer.
+   * Supported options: "murmur3" and "native".
+   * (Default = "murmur3")
+   * @group param
+   */
+  val hashAlgorithm = new Param[String](this, "hashAlgorithm", "The hash algorithm used when " +
+    "mapping term to integer. Supported options: murmur3(default) and native.",
```
Comment (sethah) [inline mllib/src/main/scala/org/apache/spark/ml/feature/HashingTF.scala]: Instead of hard coding the supported options you could use `${feature.HashingTF.supportedHashAlgorithms.mkString(",")}`. This is done in other params, like [here](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/tree/treeParams.scala#L192).

**Comment by jkbradley** [comment]
@ssimeonov Thanks for your thoughts.  I'd recommend putting design discussion like yours in the JIRA since it will get buried in the PR, whereas JIRA comments are a bit more persistent.  I'll copy your comment to the JIRA & respond there.

File: mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala (hunk 1 above)
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala]: You can't save it as a function whenever the setter is called since the ParamMap can be modified via ```Params.set```, bypassing the setHashAlgorithm method.  I'd recommend keeping ```indexOf``` (not deprecating it) and changing it to call getHashAlgorithm and compute the index based on the correct algorithm.  ```transform()``` just needs to make sure it only checks the hashAlgorithm once (rather than on each call to indexOf).

File: mllib/src/main/scala/org/apache/spark/ml/feature/HashingTF.scala [hunk 3]
```diff
@@ -63,7 +63,18 @@ class HashingTF(override val uid: String)
     "This is useful for discrete probabilistic models that model binary events rather " +
     "than integer counts")
 
-  setDefault(numFeatures -> (1 << 18), binary -> false)
+  /**
+   * The hash algorithm used when mapping term to integer.
+   * Supported options: "murmur3" and "native".
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/HashingTF.scala]: Document that "native" is the hashing method used in Spark 1.6 and earlier.

File: mllib/src/main/scala/org/apache/spark/ml/feature/HashingTF.scala [hunk 4]
```diff
@@ -63,7 +63,18 @@ class HashingTF(override val uid: String)
...
+   * (Default = "murmur3")
+   * @group param
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/HashingTF.scala]: param -> expertParam

File: mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala [hunk 5]
```diff
@@ -97,3 +126,41 @@ class HashingTF(val numFeatures: Int) extends Serializable {
...
   }
 }
+
+object HashingTF {
+
+  private[spark] val Native: String = "native"
+
+  private[spark] val Murmur3: String = "murmur3"
+
+  private[spark] val supportedHashAlgorithms = Set(Native, Murmur3)
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala]: Can you store this as an Array in case someone makes this public later on (and does not realize it needs to be Java-friendly)?

File: mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala [hunk 6]
```diff
@@ -97,3 +126,41 @@ class HashingTF(val numFeatures: Int) extends Serializable {
...
+  private val seed = 42
+
+  /**
+   * Calculate a hash code value for the term object using the native Scala implementation.
+   */
+  private[spark] def nativeHash(term: Any): Int = term.##
+
+  /**
+   * Calculate a hash code value for the term object using
+   * Austin Appleby's MurmurHash 3 algorithm (MurmurHash3_x86_32).
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala]: Put this reference in the public documentation

File: mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala [hunk 7]
```diff
@@ -97,3 +126,41 @@ class HashingTF(val numFeatures: Int) extends Serializable {
...
+      case s: Short => hashInt(s, seed)
+      case i: Int => hashInt(i, seed)
+      case l: Long => hashLong(l, seed)
+      case f: Float => hashInt(java.lang.Float.floatToIntBits(f), seed)
+      case d: Double => hashLong(java.lang.Double.doubleToLongBits(d), seed)
+      case s: String =>
+        val utf8 = UTF8String.fromString(s)
+        hashUnsafeBytes(utf8.getBaseObject, utf8.getBaseOffset, utf8.numBytes(), seed)
+      case _ => throw new SparkException("HashingTF with murmur3 algorithm does not " +
+        "support the type of input data.")
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala]: State the type of input data in the error message

File: mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala (hunk 6 above)
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala]: Document types supported in public docs as well

File: mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala [hunk 8]
```diff
@@ -97,3 +126,41 @@ class HashingTF(val numFeatures: Int) extends Serializable {
...
+object HashingTF {
+
+  private[spark] val Native: String = "native"
+
+  private[spark] val Murmur3: String = "murmur3"
+
+  private[spark] val supportedHashAlgorithms = Set(Native, Murmur3)
+
+  private val seed = 42
+
```
Comment (hujy) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala]: I think private val seed = scala.util.Random.nextInt(42) will be better.

File: mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala [hunk 9]
```diff
@@ -54,10 +60,26 @@ class HashingTF(val numFeatures: Int) extends Serializable {
...
+   * (default: murmur3)
+   */
+  @Since("2.0.0")
+  def setHashAlgorithm(value: String): this.type = {
+    hashAlgorithm = value
+    this
+  }
+
+  /**
+   * Returns the index of the input term by different hash algorithms.
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala]: "by different hash algorithms" --> "according to the hash algorithm"
```
@see [[hashAlgorithm]]
```

File: mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala [hunk 10]
```diff
@@ -96,7 +89,7 @@ class HashingTF(val numFeatures: Int) extends Serializable {
     val termFrequencies = mutable.HashMap.empty[Int, Double]
     val setTF = if (binary) (i: Int) => 1.0 else (i: Int) => termFrequencies.getOrElse(i, 0.0) + 1.0
     document.foreach { term =>
-      val i = indexOfTerm(hashAlgorithm)(term)
+      val i = indexOf(term)
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala]: This is going to check the hashAlgorithm value on every call.  I'd like us to avoid that.  If needed, you could use your indexOfTerm method but keep it private.

File: mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala [hunk 11]
```diff
@@ -66,8 +83,9 @@ class HashingTF(val numFeatures: Int) extends Serializable {
   def transform(document: Iterable[_]): Vector = {
     val termFrequencies = mutable.HashMap.empty[Int, Double]
     val setTF = if (binary) (i: Int) => 1.0 else (i: Int) => termFrequencies.getOrElse(i, 0.0) + 1.0
+    val hashFunc: Any => Int = if (hashAlgorithm == Murmur3) murmur3Hash else nativeHash
     document.foreach { term =>
-      val i = indexOf(term)
+      val i = Utils.nonNegativeMod(hashFunc(term), numFeatures)
```
Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala]: After this change, we can ensure ```transform()``` just checks the ```hashAlgorithm``` once (rather than on each call to the original ```indexOf```).

Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala]: I agree with that, but now we're back to having indexOf deprecated.  Do you mind if I push a PR to illustrate what I was thinking?

Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala]: Here's what I meant: [https://github.com/yanboliang/spark/pull/5]

Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/mllib/feature/HashingTF.scala]: @jkbradley I understand what you means now. Your PR looks great and I merged it. Thanks!

File: mllib/src/main/scala/org/apache/spark/ml/feature/HashingTF.scala [hunk 12]
```diff
@@ -31,6 +31,12 @@ import org.apache.spark.sql.types.{ArrayType, StructType}
 /**
  * :: Experimental ::
  * Maps a sequence of terms to their term frequencies using the hashing trick.
+ * Current we support two hash algorithms: "murmur3"(default) and "native".
```
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/feature/HashingTF.scala]: `Current` -> `Currently`,

and  `"murmur3"(default)` -> `"murmur3" (default)`

File: mllib/src/main/scala/org/apache/spark/ml/feature/HashingTF.scala [hunk 13]
```diff
@@ -31,6 +31,12 @@ import org.apache.spark.sql.types.{ArrayType, StructType}
...
+ * "murmur3" calculates a hash code value for the term object using
+ * Austin Appleby's MurmurHash 3 algorithm (MurmurHash3_x86_32);
+ * "native" calculates the hash code value using the native Scala implementation.
+ * In Spark 1.6 and earlier, "native" is the default hash algorithm;
+ * after Spark 2.0, we use "murmur3" as the default one.
```
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/feature/HashingTF.scala]: We must ensure to list this under the breaking changes of the 2.0 documentation

Comment (yanboliang) [inline mllib/src/main/scala/org/apache/spark/ml/feature/HashingTF.scala]: @MLnick Thanks for your comments. It has been recorded at [ML/MLlib behavior changes in 2.0](https://issues.apache.org/jira/browse/SPARK-13448) for documentation.

**Comment by jkbradley** [comment]
This LGTM.
We'll still need to update the pyspark.mllib implementation.  @yanboliang I'll merge this with master, but could you please do that in a follow-up PR?  Thanks!

**Comment by jkbradley** [comment]
Update after speaking with @mengxr 
Since we're breaking behavior anyways with with PR, perhaps we should try to enforce good practice by:
* Removing "native" support in spark.ml and pyspark.ml.
* We can leave spark.mllib as is  (no need to support it in pyspark.mllib, which would be a bit complex since we would need to create a wrapper for spark.mllib).

I'll create a JIRA for this.

**Comment by jkbradley** [comment]
Here it is: https://issues.apache.org/jira/browse/SPARK-14899
